packages/relay/tests/lib/sdkClient.spec.ts

```diff
@@ -2045,31 +2067,203 @@ describe('SdkClient', async function () {
+
+      expect(deleteFileStub.called).to.be.true;
+      hbarLimitMock.verify();
+    });
+
+    it('should execute FileInfoQuery (without paymentTransactionId) and NOT add expenses to limiter', async () => {
+      hbarLimitMock.expects('addExpense').never();
+      hbarLimitMock
+        .expects('shouldLimit')
+        .withArgs(sinon.match.any, SDKClient.queryMode, callerName)
+        .once()
+        .returns(false);
+
+      const result = await sdkClient.executeQuery(
+        new FileInfoQuery().setFileId(fileId).setQueryPayment(transactionCost),
+        client,
+        callerName,
+        interactingEntity,
+        requestId,
+      );
+
+      expect(result).to.equal(fileInfo);
+      expect(queryStub.called).to.be.true;
+      expect(queryCostStub.called).to.be.false;
+      hbarLimitMock.verify();
+    });
+
+    it('should execute FileInfoQuery (with paymentTransactionId) and add expenses to limiter', async () => {
+      hbarLimitMock.expects('addExpense').withArgs(transactionCost.toTinybars().toNumber()).once();
+      hbarLimitMock
+        .expects('shouldLimit')
+        .withArgs(sinon.match.any, SDKClient.queryMode, callerName)
+        .once()
+        .returns(false);
+
+      const result = await sdkClient.executeQuery(
+        new FileInfoQuery().setFileId(fileId).setPaymentTransactionId(transactionId),
+        client,
+        callerName,
+        interactingEntity,
+        requestId,
+      );
+
+      expect(result).to.equal(fileInfo);
+      expect(queryStub.called).to.be.true;
+      expect(queryCostStub.called).to.be.true;
+      hbarLimitMock.verify();
+    });
+
+    it('should execute EthereumTransaction and add expenses to limiter', async () => {
+      hbarLimitMock.expects('addExpense').withArgs(transactionCost.toTinybars().toNumber()).once();
+      hbarLimitMock
+        .expects('shouldLimit')
+        .withArgs(sinon.match.any, SDKClient.recordMode, callerName)
+        .twice()
+        .returns(false);
+
+      const result = await sdkClient.executeTransaction(
+        new EthereumTransaction().setCallDataFileId(fileId).setEthereumData(transactionBuffer),
+        callerName,
+        interactingEntity,
+      );
+
+      expect(result).to.equal(transactionResponse);
+      expect(transactionStub.called).to.be.true;
+      hbarLimitMock.verify();
+    });
+
+    it('should execute TransactionRecordQuery and add expenses to limiter', async () => {
+      hbarLimitMock.expects('addExpense').withArgs(transactionCost.toTinybars().toNumber()).once();
+      hbarLimitMock
+        .expects('shouldLimit')
+        .withArgs(sinon.match.any, SDKClient.recordMode, callerName)
+        .once()
+        .returns(false);
+
+      await sdkClient.executeGetTransactionRecord(
+        transactionResponse,
+        callerName,
+        interactingEntity,
+        new EthereumTransaction().constructor.name,
+      );
+
+      hbarLimitMock.verify();
+    });
```

Comment:
This is really good.
Is it possible to add one test to ensure all the above checks happen in a flow for all the transactions and queries for say a transaction than needs 2 file appends?

Reply:
Done:
- I added a new test `should rate limit before creating file and add expenses to limiter for large transaction data` which executes an `EthereumTransaction` with `callData` which needs 13 file append chunks - this is the most detailed test of all, also covers expenses for `FileContentsQuery` to retrieve the `FeeSchedules`
- I also updated the test `should execute FileCreateTransaction with callData.length > fileAppendChunkSize and add expenses to limiter` to use `callData.length := FILE_APPEND_CHUNK_SIZE * 2 + 1`